### DISCUSSION TITLE
Frigate with Coral Dual TPU and some oddities I found

### DISCUSSION THREAD

**Discussion by CancunManny** [blakeblackshear/frigate#9138] [Ask A Question]
I am very new at frigate and its toys.  I was able to get two dual coral tpu's but found out there were not adapters available.  Got a couple from China.

While I was waiting for the adapters I had frigate running cpu1 and cpu2 as detectors.  For the most part everything was working fine.  My feeds come and go, but something tells me that is more of a wifi issue as all 4 cameras currently connected are via wifi.

Well yesterday I finally got my two adapters so I started the journey to get one of them installed.  Took me a while but finally got my frigate to see and use my coral TPUs.  I assigned both of them (pci0 and pci1) and I took out the cpu detectors.

After doing so I noticed my frigate was acting up.  Half the time it would just hang, if I hit refresh it would give me connection refused messages a few times before it actually loaded.  On my homeassistant I was able to see the firgate cams intermitently.

So I went back to config and added the CPUs and took off the Corals.  And my frigate got stable again.  I then added one of the corals, with the two cpus as detectors, and it was stable.  Figured I needed the cpus as detectors for it to be stable.  I then went ahead and added the 2nd coral, and frigate got unstable again.

Figured the two cpus and one coral as detector was the correct recipe, but then I went ahead and just left the one coral tpu as the detector, and frigate is stable.

The conclusion I came up with is that Frigate didn't like it when it had two coral tpus assigned to it.  I have no idea why this is but for now that will be my setup.

If anyone is using the dual TPU and has issues consider turning one off.  If someone knows what might be causing the issue, or how to properly use both of the TPUs I would be willing to experiment

**Comment by CancunManny** [comment]
So it is about 3 hours after my original post and now I am almost speechless!  Seems I finally have exactly what I've been looking for, and good thing because I only had one more trick to try (5 port gig switch).

Before the coral TPU, I was using two cpu cores (cpu1 and cpu2).  The system worked ok, but the streams never seemed to be steady enough.  I kept trying to make excuses about it being cheap wyze cams, poor wifi connectivy, etc, but if that was true why was the fee via wyze app so much better?

Figured maybe bad router, TPLinch Archer 70, so I configured an OPNsense router. Can't say I noticed much improvement on the streams.  My TPU adapters were on their way, but I noticed I had 1 gig coming out of my OPNsense, but the switch/access point I was connecting to was only 100 on each port, including the incoming.  Figured maybe since I was chocking everything down through that slow switch my feeds were laggy, (but why not on wyze app?).  I now have a small 5 port giga switch on its way.

But yesterday my adapters came in.  And after the struggles you read on my original post finally figured out that using only ONE TPU was the key.  My NVR system is running like NEVER before.  My previous set up was with zoneminder, using same wyze cams.  Again the results for the price were "acceptable", but seemed nothing I did provided me with better streams than the ones I was getting from the app.

The way my 4 camera streams have being going the last 3 hours is amazing.  Out of the 4 cams only one drops now and then, and not only is the furthest one from the AP, it is also the one with funky test firmware which I plan to replace soon.  With my previous setup with zoneminder and home assistant, about 30% of the time I would have to hit reload a couple of times before the stream started, if it even started.  For the previous few days, with frigate without TPU and home assistant it seemed even worse than with zoneminder.  More often than not at least one cam wouldn't be streaming and usually more than one wouldn't be streaming.

With a single TPU the difference is just amazing!  The streams I am getting are actually better than the ones I get from wyze app.  With this quality of feed I am thinking of ditching the wyze app all together and putting the cams on a separate vlan with no internet.

So to finish my sharing with a question.  At the moment I have 4 wifi cams, have 3 more ready to install, and plan on getting two POE cams, for a total of 9, so lets round up to 10.  If I want to run detection in all 10 cams at about 10 FPS, would actually running a dual TPU make much of a difference?  Or for the number of cams a single TPU is more than enough?

Cause I am thinking, since I can't run both on the same frigate install, why not do another frigate install with the other half of the cams running on that one, then using the second TPU there.  Wonder if it would work as expected or if it would cause issues like it does when both run on same instance of frigate. 

In fact maybe early next week I might run two tests.  1st run a 2nd frigate and try to use the 2nd PCI there to see if both frigate run ok or if they both spaz out.  Also plan to insert both of my TPU adapters, so basically I will be running two dual Coral TPUs.   I know using two TPU from same pcie slot doesn't work, how about if i just use a single TPU from each PCIe slot?

**Comment by NickM-27** [comment]
a few things here:
1. a coral and cpu detector should never be mixed, that will just hurt performance
2. there are a number of users (including myself) that run dual tpus. If I had to guess, it might be a problem with the adapter you purchased as I use the magic blue smoke adapter and it has run stable for multiple years.
3. if you try dual tpu again actual data would be good like inference speeds, logs, etc. from the system page

**Reply by CancunManny** [reply]
Hi NickM.

Yes, I figured the adapters I am using are probably the cause of my issues.  This is part of the hardware I am using.

I got two of these https://www.mouser.mx/ProductDetail/212-G650-06076-01  M.2 Accelerator with Dual Edge; CORAL TPU DUAL EDGE TPU MODULE

One of my adapters is this one https://www.makerfabs.com/dual-edge-tpu-adapter-m2-2280-b-m-key.html  with this https://www.amazon.com.mx/dp/B084YZB5YJ?psc=1&ref=ppx_yo2ov_dt_b_product_details  (basically two adapters together)

The other adapter I got is this one https://www.makerfabs.com/dual-edge-tpu-adapter.html 

I would be happy to share data after I do more tests.  For what it is worth when I had both TPUs assigned to one frigate it was acting up but the logs didn't seem to show much info.  I am guessing there are better logs I can take a look at.

Now for up to 10 cameras, should I continue to chase my tail and try to get both of the TPUs to work, or I wouldn't see much difference if I run both at the same time, hence no need to continue wasting time.

**Reply by NickM-27** [reply]
There's no real benefit of multiple corals until you consistently see skipped detect fps.
